SSA/Projects/CIRCT/Comb/Comb.lean

```diff
@@ -0,0 +1,617 @@
+    (regArgs := .nil)
+
+def xor {Γ : Ctxt _} (arity: Nat) (a : HVector Γ.Var (List.replicate arity (.bv w))) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .xor w arity)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := a)
+    (regArgs := .nil)
+
+
+/-- Convert a list of dependent pairs into an HVector -/
+def ofList {Γ : Ctxt _} ty : (l : List ((ty : Comb.Ty) × Γ.Var ty)) → (h : l.all (·.1 = ty)) → HVector (Γ.Var) (List.replicate l.length ty)
+| [], h => .nil
+| ⟨ty', var⟩::rest, h =>
+  have hty : ty' = ty := by simp_all
+  have hrest : rest.all (·.1 = ty) := by simp_all
+  .cons (hty ▸ var) (ofList _ rest hrest)
+
+def ofList' {α : Type u} {f : α → Type v} :
+    (xs : List (Σ a, f a)) → HVector f (xs.map Sigma.fst)
+  | ⟨_, x⟩ :: xs => .cons x (ofList' xs)
+  | [] => .nil
+
+def mkExpr (Γ : Ctxt _) (opStx : MLIR.AST.Op 0) :
+    MLIR.AST.ReaderM Comb (Σ eff ty, Expr Comb Γ eff ty) := do
+  match opStx.name with
+  | op@"Comb.parity" | op@"Comb.concat" =>
+    match opStx.args with
+    | v₁Stx::[] =>
+      let ⟨ty₁, v₁⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₁Stx
+      match ty₁, op with
+      | .bv w, "Comb.parity" => return ⟨_, .bool, parity v₁⟩
+      | .hList l, "Comb.concat" => return ⟨_, .bv l.sum, concat v₁⟩
+      | _, _ => throw <| .generic s!"type mismatch"
+    | _ => throw <| .generic s!"expected one operand for `monomial`, found #'{opStx.args.length}' in '{repr opStx.args}'"
+  | op@"Comb.add" | op@"Comb.and" | op@"Comb.mul" | op@"Comb.or" | op@"Comb.xor" =>
+      let args ← opStx.args.mapM (MLIR.AST.TypedSSAVal.mkVal Γ)
+      if hl: args.length ≤ 0 then
+        throw <| .generic s!"empty list of arguments for '{repr opStx.args}'"
+      else
+        have hl' : (0 : Nat) < args.length := by
+          simp [Nat.gt_of_not_le (n := args.length) (m := 0) hl]
+        match args[0], op with
+        | ⟨.bv w, _⟩, "Comb.add" =>
+          if hall : args.all (·.1 = .bv w) then
+            let argsᵥ := ofList (.bv w) _ hall
+            have heq : args.length - 1 + 1 = args.length := by omega
+            return ⟨_, .bv w, add args.length (heq ▸ argsᵥ)⟩
+          else
+            throw <| .generic s!"Unexpected argument types for '{repr opStx.args}'"
+        | ⟨.bv w, _⟩, "Comb.and" =>
+            if hall : args.all (·.1 = .bv w) then
+              let argsᵥ := ofList (.bv w) _ hall
+              have heq : args.length - 1 + 1 = args.length := by omega
+              return ⟨_, .bv w, and args.length (heq ▸ argsᵥ)⟩
+            else
+              throw <| .generic s!"Unexpected argument types for '{repr opStx.args}'"
+        | ⟨.bv w, _⟩, "Comb.mul" =>
+            if hall : args.all (·.1 = .bv w) then
+              let argsᵥ := ofList (.bv w) _ hall
+              have heq : args.length - 1 + 1 = args.length := by omega
+              return ⟨_, .bv w, mul args.length (heq ▸ argsᵥ)⟩
+            else
+              throw <| .generic s!"Unexpected argument types for '{repr opStx.args}'"
```

Comment:
This code is highly repetitive and doing more-or-less the same thing in each branch, modulo the specific opcode emitted, there should be some function we can factor out of this to reduce the duplication.

Reply:
of course, I will extract some functions depending on the number of operands I guess